icu4c/source/i18n/ucol_sit.cpp

```diff
@@ -390,37 +390,37 @@ static void
 ucol_sit_calculateWholeLocale(CollatorSpec *s) {
     // put the locale together, unless we have a done
     // locale
-    if(s->locale[0] == 0) {
+    if(s->locale.size() == 0) {
         // first the language
-        uprv_strcat(s->locale, s->locElements[UCOL_SIT_LANGUAGE]);
+        s->locale.append(s->locElements[UCOL_SIT_LANGUAGE]);
```

Comment:
`std::string` will throw on memory allocation failure, so every single one of these `append` calls might possible throw an exception if the string class needs to reallocate and OOM happens.
AFAIK, exception usage is banned inside the ICU4C library code. Some of the tests do use the STL, and (for better or worse) don't bother checking for exceptions being thrown (which is likely okay since it's test code) -- but at least within the library code, exceptions aren't allowed.

Reply:
OK… what do you recommend to use instead?

Reply:
IIRC, I think Markus suggested to use the `CharString` class for things like this.

Here's an example usage in another i18n file:
[https://github.com/unicode-org/icu/blob/730207965353dad8856cd71ddaefbf36afd523e7/icu4c/source/i18n/dtptngen.cpp#L621](https://github.com/unicode-org/icu/blob/730207965353dad8856cd71ddaefbf36afd523e7/icu4c/source/i18n/dtptngen.cpp#L621)

It has/takes/uses an `UErrorCode` which can be use for checking OOM, etc.

It looks like this method (`ucol_sit_calculateWholeLocale`) is an internal helper, and all the callers do have existing status params -- so you could probably add an error status to the params for this helper. :)


---
In reply to: [395936645](https://github.com/unicode-org/icu/pull/1052#discussion_r395936645) [](ancestors = 395936645)

Reply:
thx i'll do that